lib/openstax/page.rb

```diff
@@ -0,0 +1,66 @@
+module Openstax
+  class Page
+    MATCH_PARAGRAPH = '//p[@id]'
+    MATCH_FIGURE = '//figure'
+    MATCH_SELECTOR_PARAGRAPH = 'p'
+    MATCH_SELECTOR_FIGURE = 'figure'
+    MATCH_FIGURE_VISIBLE_TEXT = './/figcaption'
+    MATCH_FIGURE_HIDDEN_TEXT = './/*[@data-alt]'
+
+    MATCH_ALL_ELEMENTS = [
+      MATCH_PARAGRAPH,
+      MATCH_FIGURE
+    ].join(' | ')
+
+    attr_reader :indexable_elements
+
+    def initialize(id:, data:)
+      @id = id
+      @data = data
+      @indexable_elements = []
+    end
+
+    def process_for_indexable_objects
+      content_dom.xpath(MATCH_ALL_ELEMENTS).each_with_index do | xpath_element, index |
+        indexable_element = create_element(xpath_element, index+1)
+        @indexable_elements << indexable_element if indexable_element
+      end
+    end
+
+    private
+
+    def create_element(ordered_element, page_position)
```

Comment:
jp wanted to have all this in one file, but i would prefer put the `create_[element_type]` and `matches?` 
 in separate type objects in a list list so that you can iterate over that and do `matches?` rather than having a conditional in this function for every type

Reply:
I agree with this -- the one file was only nice if it worked out this way, just because it may have been easier when we got to the 2nd and 3rd indexing strategies.